filesystem/hdfs.go

```diff
@@ -0,0 +1,165 @@
+package filesystem
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"regexp"
+
+	"github.com/colinmarc/hdfs"
+)
+
+// Requirement:
+//   hadoop version:
+//   HDFS version:
+//   REST API version:
+
+type hdfsConfig struct {
+	namenodeAddr string
+	webHdfsAddr  string
+	user         string
+}
+
+type HdfsClient struct {
+	client *hdfs.Client
+	hdfsConfig
+}
+
+func NewHdfsClient(namenodeAddr, webHdfsAddr, user string) (Client, error) {
+	client, err := hdfs.NewForUser(namenodeAddr, user)
+	if err != nil {
+		return nil, err
+	}
+	// client.Remove("/tmp/logs/testing")
+	return &HdfsClient{
+		client: client,
+		hdfsConfig: hdfsConfig{
+			namenodeAddr: namenodeAddr,
+			webHdfsAddr:  webHdfsAddr,
+			user:         user,
+		},
+	}, nil
+}
+
+func (c *HdfsClient) OpenReadCloser(name string) (io.ReadCloser, error) {
+	return c.client.Open(name)
+}
+
+func (c *HdfsClient) OpenWriteCloser(name string) (io.WriteCloser, error) {
+	exist, err := c.Exists(name)
+	if err != nil {
+		return nil, err
+	}
+	if !exist {
+		err := c.client.CreateEmptyFile(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &HdfsFile{
+		path:       name,
+		logger:     log.New(os.Stdout, "", log.Lshortfile|log.LstdFlags),
+		hdfsConfig: c.hdfsConfig,
+	}, nil
+}
+
+func (c *HdfsClient) Exists(name string) (bool, error) {
+	_, err := c.client.Stat(name)
+	return existCommon(err)
+}
+
+func (c *HdfsClient) Rename(oldpath, newpath string) error {
+	return c.client.Rename(oldpath, newpath)
+}
+
+func (c *HdfsClient) Glob(dirname, pattern string) ([]string, error) {
+	fileInfoList, err := c.client.ReadDir(dirname)
+	if err != nil {
+		return nil, err
+	}
+	res := make([]string, 0)
+	for _, fi := range fileInfoList {
+		matched, err := regexp.MatchString(pattern, fi.Name())
+		if err != nil {
+			return nil, err
+		}
+		if matched {
+			res = append(res, fi.Name())
+		}
+	}
+	return res, nil
+}
+
+type HdfsFile struct {
+	path   string
+	logger *log.Logger
+	hdfsConfig
+	// buffer
+}
+
+// REST docs:
+// http://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/WebHDFS.html#Append_to_a_File
+func (f *HdfsFile) Write(b []byte) (int, error) {
```

Comment:
So to write some data, it will be like
```
(hdfsFile, err) = hdfsClient.OpenWriteCloser(targetPath)
...
hdfsFile.Write(myBuffer)
```
It's OK to me.